File: Stack/394. Decode String/solution.cpp

class Solution {
public:
    string decodeString(string s) {
        string ss;
        int cnt = 0;
        stack<tuple<string,int>> stk;
        for (char c : s) {
            if (isdigit(c)) {
                cnt = cnt*10 + (c-'0');
            } else if (c == '[') {
                stk.push(tuple<string,int> {ss, cnt});
                ss = "";
                cnt = 0;
            } else if (c == ']') {
                auto [prefix, times] = stk.top();
                stk.pop();

                string tmp = prefix;
                for (int i=0; i<times; i++) {
                    tmp += ss;
                }
                ss = tmp;
            } else {
                ss += c;
            }
        }
        
        return ss;
    }
};

File: Stack/394. Decode String/solution.py
class Solution:
    def decodeString(self, s: str) -> str:
        count = 0
        stack = []
        cur = ''
        for ch in s:
            if ch.isdigit():
                count = count * 10 + int(ch)
            elif ch == '[':
                stack.append((cur, count))
                cur = ''
                count = 0
            elif ch == ']':
                prefix, times = stack.pop()
                cur = prefix + cur * times
            else:
                cur += ch
        return cur